### ISSUE TITLE
No Support For TLSmode require

### ISSUE THREAD

**Issue by regalme** [vertica/vertica-python#507]
Please add TLSmode support to the connection settings so that we can set TLSmode=require so that we can connect to our database with the vertica client python package.

**Comment by regalme** [comment]
> > Without this we have been unable to connect to our database using vertica_python
> 
> Have you tried to set `'ssl': True` in the connection settings?

Sure did, around 9.2 I believe is when vertica added TLSmode to ODBC, JDBC, and vsql. I have no control over our database and when they moved it to the cloud it started requiring TLSmode=require . Can absolutely not connect without that being set as an option on the client side. The python client seems to be the only client that is missing the feature.

**Comment by regalme** [comment]
> If `'ssl': True` doesn't work, you may need `'ssl': <ssl.SSLContext object>`, which gives you the full control of the ssl socket: https://github.com/vertica/vertica-python#tlsssl
> 
> 

For connections with, JDBC, ODBC, and vsql the only parameter we need to provide outside of the normal user, host, password, ect is TLSmode=require.. Any idea on what to provide on the SSLContext to mimic the same? I've looked at the decompiled JDBC driver at the TLSmode class and it's a little lost on me how to get it plugged into python.

**Comment by regalme** [comment]
> Try the following setting for your ssl.SSLContext object:
> 
> ```python
>   context.verify_mode = ssl.CERT_NONE
>   context.check_hostname = False
> ```

Still a nope.. getting vertica_python.errors.ConnectionError: [SSL: DH_KEY_TOO_SMALL] dh key too small (_ssl.c:1002)

![image](https://user-images.githubusercontent.com/29292667/235711694-c07d9730-8aee-446f-b5f4-154825f831ff.png)

**Comment by regalme** [comment]
> It is NOT
> 
> ```python
> conn_info = {
>     ...
>     'ssl': True,    # this SSL setting take effect
>     'ssl_options': {
>         'ssl_context': ssl_context
>     }
>     ...
> }
> ```
> 
> It should be
> 
> ```python
> conn_info = {
>     ...
>     'ssl': ssl_context,
>     ...
> }
> ```
> 
> I notice your error message is "dh key too small". Looks like this is not a problem of setting TLSmode to `require` or not. It is more of a system problem: https://stackoverflow.com/questions/64354210/python-dh-key-too-small-which-side-is-faulty https://stackoverflow.com/questions/61626206/what-could-cause-dh-key-too-small-error The server is offering a weak DH key, the client wants a stronger key. The problem should usually be fixed at the server side.

There is 0 issue connecting to the server with any other driver ODBC, JDBC, vsql, we supply supply TSLmode require.. But I move the context to the ssl as you suggested above and still the same error..

**Comment by regalme** [comment]
> Did you install all other drivers ODBC, JDBC, vsql on the same machine as vertica-python? What's your server version?

Yes I have used all drivers and even ODBC in python., and would love to find a work around to use vertica_python until the vertica_python driver gets updated to match what was done in all the other client drivers back at version 9.3, which was add a parameter TLSmode, which take 4 options of "disable, require, verify-ca, and verify-full. If you look at the TLSmode class in their java driver you can see what I'm talking about. 
![image](https://user-images.githubusercontent.com/29292667/235936659-f00ef1ea-b5ec-445c-82f9-e7490d7d09af.png)

**Comment by regalme** [comment]
> ```python
> ssl_context.check_hostname = False
> ssl_context.verify_mode = ssl.CERT_NONE
> ssl_context.set_ciphers('DEFAULT:!DH')
> ```

OMG thank you, I owe you a beer! And vertica if you are watching this... please just add TLSmode to the python driver! and type suggestions are you friend!
![image](https://user-images.githubusercontent.com/29292667/235939172-0004da83-b7ca-4155-9e21-9b143ab7dac5.png)

**Comment by regalme** [comment]
@sitingren  I know see you are a member... do you think you could possibly add some of that functionality under a keyword TLSmode ... so the same connection string that works for ODBC and JDBC would work for python

**Comment by KeeganCarter11** [comment]
@regalme Can you post a code snippet of what the end solution was?

**Comment by regalme** [comment]
`def ssl_context() -> ssl.SSLContext:
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_NONE
    context.set_ciphers('DEFAULT:!DH')
    return context

VERTICA_CONNECTION_INFO = {
    'host': 'nopet.com',
    'port': 5433,
    'user': hrh_user,
    'password': hrh_password,
    'database': 'HRH',
    'searchPath': hrh_schema,
    'ssl': ssl_context()
}

connect(**VERTICA_CONNECTION_INFO)

`

**Comment by regalme** [comment]
I also did this so I can get type suggestions

```
from typing import List


# Wrapper for results Row
class Row(List[any]):
    pass


# Wrapper for vertica_python.cursor.Cursor
class VerticaCursor:
    def __init__(self, cursor: any) -> None:
        self.cursor = cursor

    def execute(self, query: str) -> None:
        self.cursor.execute(query)

    def close(self) -> None:
        self.cursor.close()

    def fetchall(self) -> List[Row]:
        result = self.cursor.fetchall()
        return result


# Wrapper for vertica_python.connection.Connection
class VerticaConnection:
    def __init__(self, connection: any) -> None:
        self.connection = connection

    def cursor(self) -> VerticaCursor:
        return VerticaCursor(self.connection.cursor())

    def close(self) -> None:
        self.connection.close()

```
